### PR TITLE
feat: static hub

### PR DESCRIPTION
**Pull request by bm424**
Adds a status bar item which will open a Webview panel serving (via an iframe), the sourcery dashboard.

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -1,173 +1,283 @@
...
+import {
+  ExecuteCommandParams,
+  ExecuteCommandRequest,
+  LanguageClient,
+  LanguageClientOptions,
+  ServerOptions,
+} from "vscode-languageclient";
 
 function createLangServer(context: ExtensionContext): LanguageClient {
+  const token = workspace.getConfiguration("sourcery").get<string>("token");
```
Comment (Hellebore) [inline src/extension.ts]: let's not reformat to two spaces?

File: src/extension.ts [hunk 2]
```diff
@@ -127,6 +138,73 @@ export function activate(context: ExtensionContext) {
         languageClient.sendRequest(ExecuteCommandRequest.type, request);
     }));
 
+    // Create the "open hub" command
+    // This is activated from the status bar (see below)
+    context.subscriptions.push(
+    commands.registerCommand("sourcery.hub.open", () => {
+
+      // Command is handled by the language server
```
Comment (Hellebore) [inline src/extension.ts:146]: Change this comment here to something like - instruct the language server to start the server

File: src/extension.ts [hunk 3]
```diff
@@ -127,6 +138,73 @@ export function activate(context: ExtensionContext) {
...
+        <body style="height: 100%; width: 100%; margin: 0; padding: 0">
+          <div class="container">
+            <iframe
+              id="theFrame"
+              src="http://127.0.0.1:61816/home"
+            ></iframe>
+          </div>
+        </body>
+      </html>`;
+    })
```
Comment (Hellebore) [inline src/extension.ts:198]: this should be in a separate file

**Review by Hellebore — APPROVED** [review]
Nice and straightforward

File: src/extension.ts [hunk 4]
```diff
@@ -127,6 +138,73 @@ export function activate(context: ExtensionContext) {
...
+      const panel = window.createWebviewPanel(
+        "sourceryhub",
+        "Sourcery Hub",
+        ViewColumn.Active,
+          {
+            enableScripts: true,
+          }
+      );
+      panel.webview.html = `<!DOCTYPE html>
+        <head>
```
Comment (brendanator) [inline src/extension.ts:164]: **💄 polish** ‏Include a Sourcery favicon

File: src/extension.ts [hunk 5]
```diff
@@ -127,6 +138,73 @@ export function activate(context: ExtensionContext) {
...
+        </body>
+      </html>`;
+    })
+    );
+
+    // Create the status bar
+    const myStatusBarItem = window.createStatusBarItem(StatusBarAlignment.Left);
+    myStatusBarItem.command = "sourcery.hub.open";
+    myStatusBarItem.text = "Sourcery Hub";
+    myStatusBarItem.tooltip = "Manage your Sourcery account"
```
Comment (brendanator) [inline src/extension.ts:205]: **💄 polish** Maybe call this just `Sourcery` and use the tooltip to talk more about `settings` than account

For reference here's tabnine
image

File: src/extension.ts [hunk 6]
```diff
@@ -127,6 +138,73 @@ export function activate(context: ExtensionContext) {
...
+      // Command is handled by the language server
+      languageClient
+        .sendRequest(ExecuteCommandRequest.type, {
+          command: "sourcery.openHub",
+          arguments: [],
+        });
+
+      // Open a webview panel and fill it with a static empty page
+      // The iframe handles loading the actual content
+      const panel = window.createWebviewPanel(
```
Comment (brendanator) [inline src/extension.ts:155]: **🔨 issue:** ‏Click the hub button several times opens multiple copies of this

It'd be good to just focus if it is already open

**Review by brendanator — APPROVED** [review]
Nice!

File: src/extension.ts (hunk 4 above)
Comment (bm424) [reply src/extension.ts:164]: Done - I don't think this has any effect in VSCode though.

Comment (brendanator) [reply src/extension.ts]: Ahh, maybe there's another mechanism? Tabnine have a favicon like image in their hub

**Pull request merged**